Feat/auto completion

AHen-C0de opened the pull request:
- added component for drop down buttons
- added functions to match string of item input field with all available items in fake DB
- matched items are rendered as drop down buttons
- added useEffect to remove display of drop down buttons, when shopping list was edited
- added functionality to display drop down buttons, when focus is on item input field & to remove buttons when focus is lost (using useEffects and onFocus/onBlur)

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
+  const [allItems, setAllItems] = useState(getAllItemsFromDB);
+  const [itemInput, setItemInput] = useState("");
+  const [dropDownItems, setDropDownItems] = useState([]);
+  const [dropDownReset, setDropDownReset] = useState(false);
```
please rename to a "boolean" name: isDropDownReset for example

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
+    //filter out matched items, that are already on the shopping list
+    const uniqueMatchedItems = matchedItems.filter((matchedItem) => {
+      return (
+        items.find((item) => matchedItem.id === item.item_id) === undefined
```
instead of using this find hack you can use "indexOf"
```js
return items.indexOf(matchedItem) === -1
```
indexOf gives you the index of the first occurrence of an entity in an array. If it doesn't find any (what you want) it returns -1.

F-Kirchhoff replied:
Also you might want to filter out all used items from allItems in the first step, since then they already used items don't even match your string in the first place. 
1. reduce the set which you want to search in 
2. perform the search

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
-    onAdd(trimmedItemName);
-    form.reset();
-    inputRef.current.focus(); //set focus on input field
+  function onSubmit(event) {
+    event.preventDefault();
+    event.target.reset();
+    inputRef.current.focus();
   }
```
this only resets the form, is this intended?

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
-      <StyledForm aria-label="add items" onSubmit={submitForm}>
+      <StyledForm
+        aria-label="add items"
+        onSubmit={(event) => onSubmit(event)}
```
if you want, you can write this shorter:
```
onSubmit={onSubmit}
```

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
-          ref={inputRef} // set ref to set autofocus after submit
+          ref={inputRef} //set ref to set autofocus after submit
+          value={itemInput}
+          onInput={(event) => setItemInput(event.target.value)} //don't use onChange() -> it ignores some events!!!
+          onFocus={triggerDropDown}
+          onBlur={() => setDropDownReset(true)} //to close drop down, when losing focus
```
There is a simpler way of displaying the dropdown. use a state with all the dropdownitems, starting with an empty array. Then use conditional rendering which disables the dropdown if the array is empty (something like 
`{dropdownItems.length <=0 && ...}`. Then search for matching items every time the input value changes. On a click on the add button of an dropDownItem you could either clear the form input which or set the dropDownItems to an empty array, both resulting in hiding the dropdown list.

F-Kirchhoff replied:
You still need the onBlur function, but here you can simply reset the input and clear the dropdownItems array

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
-    const formData = new FormData(form);
-    const itemName = Object.fromEntries(formData).item;
+  //open drop down when typing into input field
+  useEffect(() => {
+    triggerDropDown();
+  }, [itemInput]);
```
by using useEffect you double the rerenders for each change in the input. Why not calling triggerDropDown directly in the onChange function of the input?

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
-    const trimmedItemName = itemName.trim(); //remove white spaces at beginning and end of the name
+  //reset dropDownItems state, to remove dropDown buttons from screen,...
+  //...when shopping list (items) changed
+  useEffect(() => {
+    setItemInput("");
+    setDropDownItems([]);
+  }, [items]);
```
not necessary if you apply the method i described below

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
     }
-    if (
-      items.find(
-        (item) => item.name.toLowerCase() === trimmedItemName.toLowerCase()
-      ) !== undefined
-    ) {
-      alert("Item already on the list!");
-      return;
+  }, [dropDownReset]);
```
not necessary if you apply the method i described below

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -2,55 +2,99 @@ import styled from "styled-components";
+
+  //match input with all available items from DB
+  function matchInput(value) {
+    const editedValue = value.trim().toLowerCase(); //trim() -> remove white spaces at beginning and end of the name
+    //clear drop down when clearing input field
+    if (editedValue === "") {
+      return [];
     }
```
nice!

F-Kirchhoff reviewed (changes requested):
Hey, great work already. I added some comments how to improve the code. In general you can apply the following rule of thumb: Do I really need a useEffect here? If not, solve the problem differently. A useEffect is only necessary if you want to apply "side Effects" (fetching, localstorage, adding event listeners to the document / window, hydrating a state)

AHen-C0de commented:
> Hey, great work already. I added some comments how to improve the code. In general you can apply the following rule of thumb: Do I really need a useEffect here? If not, solve the problem differently. A useEffect is only necessary if you want to apply "side Effects" (fetching, localstorage, adding event listeners to the document / window, hydrating a state)

Ah ok, I didn't know that! So as you suggested, I removed all of these useEffects and exchanged them with functions. Also, I exchanged the `onClick` in my DropDown component with a `onMouseDown` to trigger the click BEFORE the `onBlur` event of the input field, which removes the drop down

F-Kirchhoff commented on components/ShoppingListEditor/InputDropDown.js:
```diff
@@ -1,11 +1,24 @@
+  onReset,
+}) {
+  function handleClick(element) {
+    onAdd(element);
+    onReset();
```
Why not combining both into one function? Or do you call one of these functions independently from another?

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -9,7 +9,7 @@ export default function ShoppingListEditor({ items, onDelete, onAdd }) {
   const [itemInput, setItemInput] = useState("");
   const [dropDownItems, setDropDownItems] = useState([]);
-  const [dropDownReset, setDropDownReset] = useState(false);
+  const [isFocusInput, setIsFocusInput] = useState(false);
```
Ahh, i love the naming here ❤️

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -18,64 +18,63 @@ export default function ShoppingListEditor({ items, onDelete, onAdd }) {
-    const matchedItems = allItems.filter((item) =>
+
+    //filter out items from allItems, that are already on the shopping list
+    const usedItemIds = items.map((usedItem) => usedItem.item_id);
+    const availableItems = allItems.filter(
+      (DBitem) => !usedItemIds.includes(DBitem.id)
+    );
```
nice!

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -86,22 +85,28 @@ export default function ShoppingListEditor({ items, onDelete, onAdd }) {
-          ariaLabel="add item"
-          onAdd={onAdd}
+          onInput={(event) => handleItemInput(event)} //don't use onChange() -> it ignores some events!!!
+          onFocus={() => triggerDropDown(itemInput)}
+          //close drop down, when losing focus
```
what exactly happens here if I focus the input but don't type anything yet? I guess the dropdown won't be displayed since the dropDownItems.length is still 0. I guess you don't even need to check if the input is focussed, since the dropdownItems are displayed as soon as the array gets some elements. This happens when you write something in the input, for that the input needs to be focused anyway. Check if it still works when you leave out this event listener.

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -18,64 +18,63 @@ export default function ShoppingListEditor({ items, onDelete, onAdd }) {
+
+  function handleBlur() {
+    setDropDownItems([]);
+    if (isFocusInput) {
+      inputRef.current.focus();
+      setIsFocusInput(false);
+    }
```
what does this part do? If I remove the focus from the input I focus it again?

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -18,64 +18,63 @@ export default function ShoppingListEditor({ items, onDelete, onAdd }) {
-    event.preventDefault();
-    event.target.reset();
-    inputRef.current.focus();
+  function onAddReset() {
+    setItemInput("");
+    setDropDownItems([]);
+    setIsFocusInput(true); //to focus item input after adding item to list
+  }
```
As mentioned above, you could write a function handleDropDownAdd which looks like this:
```js
function handleDropDownAdd(element) {
onAdd(element) // the function from the props
 setItemInput("");
 setDropDownItems([]);
 setIsFocusInput(true); 
}



```

F-Kirchhoff commented on components/ShoppingListEditor/ShoppingListEditor.js:
```diff
@@ -86,22 +85,28 @@ export default function ShoppingListEditor({ items, onDelete, onAdd }) {
+          <InputDropDown
+            optionElements={dropDownItems}
+            ariaLabel="add item"
+            onAdd={onAdd}
+            onReset={onAddReset}
```
here you would hand down only one function

F-Kirchhoff reviewed (changes requested):
Hey, this is looking much better, still found something to improve though. Have a look!

F-Kirchhoff commented:
LGTM 🦊
